handler/rfc8693/refresh_token_type_handler.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	form := request.GetRequestForm()
+	requestedTokenType := form.Get("requested_token_type")
+	if requestedTokenType == "" {
+		requestedTokenType = c.Config.GetDefaultRequestedTokenType(ctx)
+	}
+
+	if requestedTokenType != RefreshTokenType {
+		return nil
+	}
+
+	if err := c.issue(ctx, request, responder); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CanSkipClientAuth indicates if client auth can be skipped
+func (c *RefreshTokenTypeHandler) CanSkipClientAuth(ctx context.Context, requester fosite.AccessRequester) bool {
+	return false
+}
+
+// CanHandleTokenEndpointRequest indicates if the token endpoint request can be handled
+func (c *RefreshTokenTypeHandler) CanHandleTokenEndpointRequest(ctx context.Context, requester fosite.AccessRequester) bool {
+	// grant_type REQUIRED.
+	// Value MUST be set to "password".
+	return requester.GetGrantTypes().ExactOne("urn:ietf:params:oauth:grant-type:token-exchange")
+}
+
+func (c *RefreshTokenTypeHandler) validate(ctx context.Context, request fosite.AccessRequester, token string) (
+	fosite.Session, map[string]interface{}, error) {
+
+	session, _ := request.GetSession().(Session)
+	if session == nil {
+		return nil, nil, errorsx.WithStack(fosite.ErrServerError.WithDebug(
+			"Failed to perform token exchange because the session is not of the right type."))
+	}
+
+	client := request.GetClient()
+
+	sig := c.CoreStrategy.RefreshTokenSignature(ctx, token)
+	or, err := c.Storage.GetRefreshTokenSession(ctx, sig, request.GetSession())
+	if err != nil {
+		return nil, nil, errors.WithStack(fosite.ErrInvalidRequest.WithHint("Token is not valid or has expired.").WithDebug(err.Error()))
+	} else if err := c.CoreStrategy.ValidateRefreshToken(ctx, or, token); err != nil {
+		return nil, nil, err
+	}
+
+	tokenClientID := or.GetClient().GetID()
+	// forbid original subjects client to exchange its own token
+	if client.GetID() == tokenClientID {
+		return nil, nil, errors.WithStack(
+			fosite.ErrRequestForbidden.WithHint("Clients are not allowed to perform a token exchange on their own tokens."))
+	}
+
+	// Check if the client is allowed to exchange this token
+	if subjectTokenClient, ok := or.GetClient().(Client); ok {
+		allowed := subjectTokenClient.TokenExchangeAllowed(client)
+		if !allowed {
+			return nil, nil, errors.WithStack(fosite.ErrRequestForbidden.WithHintf(
+				"The OAuth 2.0 client is not permitted to exchange a subject token issued to client %s", tokenClientID))
+		}
+	}
+
+	// Scope check
+	for _, scope := range request.GetRequestedScopes() {
+		if !c.ScopeStrategy(or.GetGrantedScopes(), scope) {
+			return nil, nil, errors.WithStack(fosite.ErrInvalidScope.WithHintf("The subject token is not granted \"%s\" and so this scope cannot be requested.", scope))
+		}
+	}
+
+	// Convert to flat session with only access token claims
+	tokenObject := session.AccessTokenClaimsMap()
```

Comment:
there's a bug here that we fix in this PR https://github.ibm.com/strata/ristretto-core/pull/253/files (should take from original, not the session)